### ISSUE TITLE
Unreasonable high GPU usage when using DogHardNet

### ISSUE THREAD

**Issue by erdemkkok** [kornia/kornia#3095]
### Describe the bug

 . GPU mem leak 160MB to 13 GB .How can i fix it ?

### Reproduction steps

```bash
1
```

### Expected behavior

less GPU mem usage

### Environment

```shell
when running LightGlue DogHardNet
- 2.5.1
-  Linux
- installed with pip

- Python version: 3.10 ~ 3.8
- CUDA/cuDNN version: 12.4
```

**Comment by edgarriba** [comment]
Can you attach a script to reproduce? /cc @ducha-aiki

**Comment by erdemkkok** [comment]
https://github.com/cvg/LightGlue/blob/main/lightglue/dog_hardnet.py when i use that repo line 38 has memory leak

Line 38 use kornia/feature/integrated.py -> return get_laf_descriptors(img, lafs, self.descriptor, self.patch_size, self.grayscale_descriptor)

**Comment by ducha-aiki** [comment]
@erdemkkok are you sure it leaks memory and just not peaks its usage? Is the memory released back after `torch.cuda.empty_cache()`?

The ways HardNet (and other patch descriptors) are extracted, is to first sample the patches, which would be [B, N, 32, 32], e.g. [1, 4000, 32, 32], and then  describe it, which would be equal to run descriptor with batch size of 4000. This takes a lot of memory. 

So, I would recommend to use extraction with `torch.inference_mode()`, as well as float16.

**Comment by erdemkkok** [comment]
can u share code example because i tried " torch.cuda.empty_cache() " and float16 convert method . if i dont use thats .MEM usage is 26 GB.

**Comment by ducha-aiki** [comment]
How many keypoints do you have?
@erdemkkok

**Comment by erdemkkok** [comment]
8000 keypoints, but matched keypoints 200

**Comment by ducha-aiki** [comment]
@erdemkkok it doesn't matter how many matched, 8000 descriptors extracted at once is a lot, so the memory consumption is what it is. 

I can make a PR, which adds a possibility to extract patches in micro-batches to LAFDescriptor, thus saving memory, but that would likely be not today.

**Comment by Mertalper24** [comment]
@ducha-aiki I am having the same issue with @erdemkkok because I am using 8192 max_num_keypoints as well cause doghardnet gives the best performance at 8192, when I lower it, the matched keypoints decreases significantly. So its actually a trade off between performance, accuracy, and memory. 

@ducha-aiki > I can make a PR, which adds a possibility to extract patches in micro-batches to LAFDescriptor, thus saving memory, but that would likely be not today.

I actually thought of the same thing and came up with something like this. Here it is.

in  lightglue/dog_hardnet.py

instead of the line
p["descriptors"] = self.laf_desc(img[None], lafs).reshape(-1, 128)

I added this
            batch_size = 512
            descriptors = []
            for i in range(0, lafs.shape[1], batch_size):
                batch_lafs = lafs[:, i:i + batch_size]
                batch_descriptors = self.laf_desc(img[None], batch_lafs).reshape(-1, 128)
                descriptors.append(batch_descriptors)
            p["descriptors"] = torch.cat(descriptors, dim=0)
            
which limits the memory allocation but slows it down of course.
Did you think of the same thing or it was something different?

**Comment by ducha-aiki** [comment]
@Mertalper24 yes, like this. If you can make PR, I will review and merge quickly

**Comment by erdemkkok** [comment]
I tried this, @Mertalper24, but the system became very slow. Do you have a solution for this, @ducha-aiki ?

**Comment by ducha-aiki** [comment]
@erdemkkok it seems that it would be better for you to use some other descriptors then. I don't know any better way.
 DoGHardNet have benefits of the scale and rotation invariance, but comes at cost of the the speed/memory.

One thing, which is not specific to DogHardNet though, is that it is much more efficient to extract all the features first, and then do matching step, this way you don't extract features from the same image multiple times. Check this notebook https://github.com/ducha-aiki/imc2023-kornia-starter-pack/blob/main/keynetaffnet-adalam-pycolmap-3dreconstruction.ipynb for an example.

**Comment by edgarriba** [comment]
feel free to re-open of there's still any interest in this. Thanks !
